File: 22_专题/离线查询/莫队/MoAlgo.py
from typing import Generic, List, TypeVar
from math import sqrt


V = TypeVar("V")  # 区间元素类型
Q = TypeVar("Q")  # 每个查询的返回值类型


class MoAlgo(Generic[V, Q]):

    """
    莫队算法模板
    左端点分桶，右端点排序
    """

    def __init__(self, n: int, q: int):
        self._chunkSize = max(1, n // int(sqrt(q)))
        self._buckets = [[] for _ in range(n // self._chunkSize + 1)]
        self._queryOrder = 0

    def addQuery(self, left: int, right: int) -> None:
        """0 <= left <= right < n"""
        index = left // self._chunkSize
        self._buckets[index].append((self._queryOrder, left, right + 1))  # 注意这里的 right+1
        self._queryOrder += 1

    def work(self) -> List[Q]:
        """返回每个查询的结果"""
        buckets = self._buckets
        res: List[Q] = [None] * self._queryOrder  # type: ignore
        left, right = 0, 0

        for i, bucket in enumerate(buckets):
            bucket.sort(key=lambda x: x[2], reverse=not not i & 1)

            for qi, qLeft, qRight in bucket:
                # !窗口扩张
                while left > qLeft:
                    left -= 1
                    self._add(left, -1)
                while right < qRight:
                    self._add(right, 1)
                    right += 1

                # !窗口收缩
                while left < qLeft:
                    self._remove(left, 1)
                    left += 1
                while right > qRight:
                    right -= 1
                    self._remove(right, -1)

                res[qi] = self._query(qLeft, qRight - 1)

        return res

    def _add(self, index: int, delta: int) -> None:
        """将数据添加到窗口"""
        raise NotImplementedError(f"{self.__class__.__name__}._add")

    def _remove(self, index: int, delta: int) -> None:
        """将数据从窗口中移除"""
        raise NotImplementedError(f"{self.__class__.__name__}._remove")

    def _query(self, qLeft: int, qRight: int) -> Q:
        """更新当前窗口的查询结果"""
        raise NotImplementedError(f"{self.__class__.__name__}._query")

File: 算法竞赛进阶指南/GoDS (Go Data Structures)/src/moalgo/moalgo/moalgo.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
)

// G - Range Pairing Query
// https://atcoder.jp/contests/abc242/tasks/abc242_g
func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	const N int = 1e5 + 10
	var n int
	fmt.Fscan(in, &n)
	nums := make([]int, n)
	for i := range nums {
		fmt.Fscan(in, &nums[i])
	}

	var q int
	fmt.Fscan(in, &q)

	pair := 0
	counter := [N + 1]int{}
	mo := NewMoAlgo(n, q, op{
		add: func(i, _ int) {
			v := nums[i]
			pair -= counter[v] / 2
			counter[v]++
			pair += counter[v] / 2
		},
		remove: func(i, _ int) {
			v := nums[i]
			pair -= counter[v] / 2
			counter[v]--
			pair += counter[v] / 2
		},
		query: func(qLeft, qRight int) int { return pair },
	})

	for ; q > 0; q-- {
		var l, r int
		fmt.Fscan(in, &l, &r)
		l--
		r--
		mo.AddQuery(l, r)
	}

	res := mo.Work()
	for _, v := range res {
		fmt.Fprintln(out, v)
	}
}

type R = int

type MoAlgo struct {
	queryOrder int
	chunkSize  int
	buckets    [][]query
	op         op
}

type query struct{ qi, left, right int }

type op struct {
	// 将数据添加到窗口
	add func(index, delta int)
	// 将数据从窗口中移除
	remove func(index, delta int)
	// 更新当前窗口的查询结果
	query func(qLeft, qRight int) R
}

func NewMoAlgo(n, q int, op op) *MoAlgo {
	chunkSize := max(1, n/int(math.Sqrt(float64(q))))
	buckets := make([][]query, n/chunkSize+1)
	return &MoAlgo{chunkSize: chunkSize, buckets: buckets, op: op}
}

// 0 <= left <= right < n
func (mo *MoAlgo) AddQuery(left, right int) {
	index := left / mo.chunkSize
	mo.buckets[index] = append(mo.buckets[index], query{mo.queryOrder, left, right + 1})
	mo.queryOrder++
}

// 返回每个查询的结果
func (mo *MoAlgo) Work() []R {
	buckets, q := mo.buckets, mo.queryOrder
	res := make([]R, q)
	left, right := 0, 0

	for i, bucket := range buckets {
		if i&1 == 1 {
			sort.Slice(bucket, func(i, j int) bool { return bucket[i].right > bucket[j].right })
		} else {
			sort.Slice(bucket, func(i, j int) bool { return bucket[i].right < bucket[j].right })
		}

		for _, q := range bucket {
			// !窗口扩张
			for left > q.left {
				left--
				mo.op.add(left, -1)
			}
			for right < q.right {
				mo.op.add(right, 1)
				right++
			}

			// !窗口收缩
			for left < q.left {
				mo.op.remove(left, 1)
				left++
			}
			for right > q.right {
				right--
				mo.op.remove(right, -1)
			}

			res[q.qi] = mo.op.query(q.left, q.right-1)
		}
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
